R/formatOccData.r

```diff
@@ -48,38 +50,96 @@
 #' site <- sample(paste('A', 1:nSites, sep=''), size = n, TRUE)
 #' 
 #' # the date of visit is selected at random from those created earlier
-#' time_period <- sample(rDates, size = n, TRUE)
+#' survey <- sample(rDates, size = n, TRUE)
 #'
 #' # run the model with these data for one species
 #' formatted_data <- formatOccData(taxa = taxa,
 #'                                 site = site,
-#'                                 time_period = time_period)
+#'                                 survey = survey,
+#'                                 includeJDay = TRUE)
+#' }
+#' @examples
+#' \dontrun{
+#' 
+#' # Create data with coarser survey information
+#' n <- 1500 #number of species observation in dataset
+#' np <- 10 # number of closure periods in data
+#' nSurveys <- 100 # set number of surveys
+#' nSites <- 20 # set number of sites
+#' 
+#' # taxa are set as random letters
+#' taxa <- sample(letters, size = n, TRUE)
+#' 
+#' # three sites are visited randomly
+#' site <- sample(paste('A', 1:nSites, sep=''), size = n, TRUE)
+#' 
+#' # the date of visit is selected at random from those created earlier
+#' survey <- sample(nSurveys, size = n, TRUE)
+#'
+#' # allocate the surveys randomly to closure periods 
+#' cp <- sample(1:np, nSurveys, TRUE)
+#' closure_period <- cp[survey]
+#'
+#' # run the model with these data for one species
+#' formatted_data <- formatOccData(taxa = taxa,
+#'                                 site = site,
+#'                                 survey = survey,
+#'                                 closure_period = closure_period)
 #' }
 #' @export
 #' @importFrom dplyr distinct
+#' @importFrom dplyr summarise
+#' @importFrom dplyr group_by
 #' @importFrom reshape2 dcast
 
-formatOccData <- function(taxa, site, time_period, includeJDay = FALSE){
+formatOccData <- function(taxa, site, survey, closure_period = NULL, includeJDay = FALSE){
 
   # Do error checks
-  errorChecks(taxa = taxa, site = site, time_period = time_period)
+  errorChecks(taxa = taxa, site = site, survey = survey, closure_period = closure_period)
+
+  # Additional error check whether survey is a date
+  if(!'POSIXct' %in% class(survey) & !'Date' %in% class(survey)){
+    # This is necessary if either closure_period is not supplied or if includeJDay = TRUE
+    if(is.null(closure_period)){
+      stop('survey must be a date if closure_period not supplied')
+    } else {
+      if(includeJDay == TRUE){
+        stop('survey must be a date if Julian Date is to be included')        
+    }
+   }
+  } # otherwise survey can be anything (including a character vector)
   
   # Create dataframe from vectors
-  taxa_data <- distinct(data.frame(taxa, site, time_period))
+  taxa_data <- data.frame(taxa, site, survey)
```

Comment:
@drnickisaac you have removed BOTH the `distinct` here AND the `unique` further down. In doing so you have introduced an error. This has been caught in the testing code and resulted in this error in on the build server.

```
── 1. Failure: Test formatOccData (@testformatOccData.r#74)  ───────────────────
head(visitData$occDetdata) not identical to `head_occDetdata`.
Component "site": Attributes: < Component "levels": Lengths (50, 32) differ (string compare on first 32) >
Component "site": Attributes: < Component "levels": 27 string mismatches >
```

It is easy to see this error if we run your example

```r
set.seed(111)

# Create data
n <- 15000 #size of dataset
nyr <- 20 # number of years in data
nSurveys <- 100 # set number of dates
nSites <- 50 # set number of sites

# Create somes dates
first <- as.Date(strptime("2010/01/01", "%Y/%m/%d"))
last <- as.Date(strptime(paste(2010+(nyr-1),"/12/31", sep=''), "%Y/%m/%d"))
dt <- last-first
rDates <- first + (runif(nSurveys)*dt)

# taxa are set as random letters
taxa <- sample(letters, size = n, TRUE)

# three sites are visited randomly
site <- sample(paste('A', 1:nSites, sep=''), size = n, TRUE)

# the date of visit is selected at random from those created earlier
survey <- sample(rDates, size = n, TRUE)

# run the model with these data for one species
formatted_data <- formatOccData(taxa = taxa,
                                site = site,
                                survey = survey,
                                includeJDay = TRUE)

head(formatted_data$occDetdata)
```

```
          visit site L   TP Jul_date
1 A102010-01-08  A10 6 2010     -174
2 A102010-01-08  A10 6 2010     -174
3 A102010-01-08  A10 6 2010     -174
4 A102010-01-08  A10 6 2010     -174
5 A102010-01-08  A10 6 2010     -174
6 A102010-01-08  A10 6 2010     -174
```

You can see that the visit is returned in multiple rows by error. This is caught by the error checks because in the testing we know what should be returned from our test data. This should be an easy fix by either adding the `distinct` or `unique`.

Reply:
I've moved the call to `distinct()` further down. But there's an additional call to `unique()`. 
It's logical to remove duplicates only once, I think. This would be at the end.

Reply:
Now we realise that `occDetdata `is created from `taxa_data`. I had assumed it was a selection of rows from `spp_vis`. We've added back the calls to `unique()`